src/v/kafka/server/group_metadata_migration.cc

```diff
@@ -0,0 +1,691 @@
+
+    co_await feature_table().await_feature(
+      cluster::feature::consumer_offsets, abort_source());
+    co_await _group_router.invoke_on_all([](group_router& router) {
+        return router.get_group_manager().local().reload_groups();
+    });
+    vlog(mlog.info, "consumer offset feature enabled");
+    co_await _group_router.invoke_on_all(&group_router::enable);
+    co_await _controller.get_topics_frontend().invoke_on_all(
+      &cluster::topics_frontend::enable_partition_movement);
+}
+
+ss::future<> group_metadata_migration::migrate_metadata() {
+    auto& topics = _controller.get_topics_state().local();
+    auto group_topic_assignment = topics.get_topic_assignments(
+      model::kafka_group_nt);
+    // no group topic found, skip migration
+    if (!group_topic_assignment) {
+        co_return;
+    }
+    auto partitions = group_topic_assignment->size();
+
+    // create consumer offsets topic based on the old configuration
+    while (!topics.contains(
+      model::kafka_consumer_offsets_nt, model::partition_id{0})) {
+        if (_controller.is_raft0_leader()) {
+            vlog(mlog.info, "creating consumer offsets topic");
+            co_await create_consumer_offsets_topic(
+              _controller,
+              std::move(*group_topic_assignment),
+              default_deadline());
+            continue;
+        }
+
+        co_await ss::sleep_abortable(
+          default_timeout, _controller.get_abort_source().local());
+    }
+
+    vlog(mlog.info, "waiting for preparing barrier");
+    co_await feature_manager().barrier(preparing_barrier_tag);
+
+    absl::node_hash_map<ss::shard_id, std::vector<model::ntp>> shard_ntps;
+    for (int16_t p_id = 0; p_id < static_cast<int16_t>(partitions); ++p_id) {
+        model::ntp ntp(
+          model::kafka_group_nt.ns,
+          model::kafka_group_nt.tp,
+          model::partition_id(p_id));
+
+        auto shard_id = _controller.get_shard_table().local().shard_for(ntp);
+
+        if (shard_id) {
+            shard_ntps[*shard_id].push_back(std::move(ntp));
+        }
+    }
+
+    for (auto& [shard, ntps] : shard_ntps) {
+        ssx::spawn_with_gate(
+          _partitions_gate, [this, shard = shard, ntps = ntps]() mutable {
+              vlog(
+                mlog.info,
+                "dispatching {} ntps migration on shard: {}",
+                ntps.size(),
+                shard);
+              return ss::smp::submit_to(
+                shard,
+                [&pm = _controller.get_partition_manager(),
+                 &st = _controller.get_shard_table(),
+                 &as = _controller.get_abort_source(),
+                 ntps = std::move(ntps)]() mutable -> ss::future<> {
+                    return dispatch_ntps_migration(std::move(ntps), pm, st, as);
+                });
+          });
+    }
+
+    co_await _partitions_gate.close();
+
+    vlog(mlog.info, "finished migrating kafka_internal/group partitions");
```

Comment:
it sorta seems like there needs to be a post migration verification that it succeeded. for example, in `do_dispatch_ntp_migration` it looks like the migration returns error free in various cases like:

```
    auto target_shard = st.local().shard_for(target_ntp);
     if (!target_shard) {
         vlog(mlog.error, "unable to find shard for {}", target_ntp);
         co_return;
     }
```

this is fine assuming that there isn't a mistake and maybe disabling partition movement had an issue or some other thing. after migration returns verifying it some how might be useful.

but I dunno. just a thought.